frontend/src/metabase/public/containers/PublicQuestion/PublicQuestion.tsx

```diff
@@ -0,0 +1,245 @@
+
+    const parameters = card.parameters || getParametersFromCard(card);
+
+    try {
+      setResult(null);
+
+      let newResult;
+      if (token) {
+        // embeds apply parameter values server-side
+        newResult = await maybeUsePivotEndpoint(
+          EmbedApi.cardQuery,
+          card,
+        )({
+          token,
+          ...getParameterValuesBySlug(parameters, parameterValues),
+        });
+      } else if (uuid) {
+        // public links currently apply parameters client-side
+        const datasetQuery = applyParameters(card, parameters, parameterValues);
+        newResult = await maybeUsePivotEndpoint(
+          PublicApi.cardQuery,
+          card,
+        )({
+          uuid,
+          parameters: JSON.stringify(datasetQuery.parameters),
+        });
+      } else {
+        throw { status: 404 };
+      }
+
+      setResult(newResult);
+    } catch (error) {
+      console.error("error", error);
+      dispatch(setErrorPage(error));
+    }
+  }, [card, dispatch, parameterValues, token, uuid]);
+
+  useEffect(() => {
+    run();
+  }, [run]);
+
+  useEffect(() => {
+    if (result && !initialized) {
+      setInitialized(true);
+    }
+  }, [initialized, result]);
+
+  const getParameters = () => {
+    if (!initialized || !card) {
+      return [];
+    }
+
+    return getCardUiParameters(
+      card,
+      metadata,
+      {},
+      card.parameters || undefined,
+    );
+  };
+
+  const question = new Question(card, metadata);
+
+  const actionButtons = result && (
+    <QueryDownloadWidget
+      className={cx(CS.m1, CS.textMediumHover)}
+      question={question}
+      result={result}
+      uuid={uuid}
+      token={token}
+    />
+  );
+
+  const { bordered, hide_download_button, hide_parameters, theme, titled } =
+    useEmbedFrameOptions({ location });
+
+  return (
+    <EmbedFrame
+      name={card && card.name}
+      description={card && card.description}
+      actionButtons={actionButtons}
+      question={question}
+      parameters={getParameters()}
+      parameterValues={parameterValues}
+      setParameterValue={setParameterValue}
+      enableParameterRequiredBehavior
+      setParameterValueToDefault={setParameterValueToDefault}
+      bordered={bordered}
+      hide_download_button={hide_download_button}
+      hide_parameters={hide_parameters}
+      theme={theme}
+      titled={titled}
+    >
+      <LoadingAndErrorWrapper
+        className={CS.flexFull}
+        loading={!result || !initialized}
+        error={typeof result === "string" ? result : null}
+        noWrapper
+      >
+        {() => (
+          <Visualization
+            error={result && result.error}
+            rawSeries={[{ card: card, data: result && result.data }]}
+            className={cx(CS.full, CS.flexFull, CS.z1)}
+            onUpdateVisualizationSettings={(
+              settings: VisualizationSettings,
+            ) => {
+              setCard(
+                updateIn(
+                  card,
```

Comment:
we should use a function updater here too, so it's consistent with the rest of the codebase.

```ts
setCard(card => xxx)
```

Reply:
fantastic catch